### PR TITLE
GH-1113 Refactor widget.data/column

### PR DESCRIPTION
**Pull request by GordonSmith**
Add Database File with a Grid Class.

Fixes GH-1113

Signed-off-by: Gordon Smith <gordon.smith@lexisnexis.com>

### CODE REVIEW

**Comment by GordonSmith** [comment]
@jbrundage @mlzummo @dtsnell4 Please Review
The new DB is used in every widget now, but here is my main test page:  http://rawgit.com/GordonSmith/Visualization/GH-1113/demos/DataAnalyst/index.html

File: src/chart/Column.js [hunk 1]
```diff
@@ -92,7 +92,6 @@
 
                 if (isHorizontal) {
                     columnRect.transition()
-                        .attr("class", "columnRect")
```
Comment (mlzummo) [inline src/chart/Column.js]: why did you remove the class?

File: src/common/Database.js [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+    if (typeof define === "function" && define.amd) {
+        define(["d3"], factory);
+    } else {
+        root.common_Database = factory(root.d3);
+    }
+}(this, function (d3) {
+    function Grid() {
+        this.clear();
+    }
+    Grid.prototype.constructor = Grid;
```
Comment (mlzummo) [inline src/common/Database.js]: this isnt in reference to the layout grid we have ... should we use the term DataMatrix? (this is my first pass through of this code so my questions might not make sense)

File: src/common/Surface.js [hunk 3]
```diff
@@ -264,7 +264,9 @@
             ;
         }
 
-        this._menu.element().node().parentNode.appendChild(this._menu.element().node()); // Make sure menu is on top (Z-Order POV)
+        if (this._menu.element() && this._menu.element().node() && this._menu.element().node().parentNode) {
```
Comment (mlzummo) [inline src/common/Surface.js]: this is to just check th menu exists and is a child before doing the Z order fix?

File: src/common/Database.js [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+                    case "SUM":
+                        return d3.sum(leaves, function (d) { return d[rollupValueIdx[0]]; });
+                    case "AVE":
+                        return d3.mean(leaves, function (d) { return d[rollupValueIdx[0]]; });
+                    case "MIN":
+                        return d3.min(leaves, function (d) { return d[rollupValueIdx[0]]; });
+                    case "MAX":
+                        return d3.max(leaves, function (d) { return d[rollupValueIdx[0]]; });
+                }
+                return "???";
```
Comment (mlzummo) [inline src/common/Database.js]: ??? .. whats that?

Comment (GordonSmith) [inline src/common/Database.js]: Should be console warning I suspect.

File: src/common/Database.js [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+            var mapping = columns[rollupField];
+            var params = [];
+            for (var param in mapping.params) {
+                params.push(mapping.params[param]);
+            }
+            var nested = this.rollup(rollupBy, function (leaves) {
+                switch (mapping.function) {
+                    case "SUM":
+                        return d3.sum(leaves, function (d) { return d[rollupValueIdx[0]]; });
+                    case "AVE":
```
Comment (dtsnell4) [inline src/common/Database.js]: Should this be "AVG"?

Comment (GordonSmith) [inline src/common/Database.js]: Alas no (and MEAN would have been my preference) - its from the DDL.

File: src/common/Surface.js (hunk 3 above)
Comment (GordonSmith) [inline src/common/Surface.js]: Yea I hot a scenario where there was no menu and just wrapped it in the if.

File: src/common/Database.js (hunk 2 above)
Comment (GordonSmith) [inline src/common/Database.js]: You will only ever see in this context:  ```var dataGrid = new Database.Grid();``` so I think its ok.

File: src/chart/Column.js (hunk 1 above)
Comment (GordonSmith) [inline src/chart/Column.js]: Its set on the enter() method above so is not needed.

**Comment by dtsnell4** [comment]
This all looks good to me.  Looking forward to trying it out.

**Comment by jbrundage** [comment]
Yes, looks good to me too

**Comment by mlzummo** [comment]
im ok with merging it

**Comment by GordonSmith** [comment]
Fixed the "???" Issue
